src/NServiceBus.RabbitMQ/MessagePump.cs

```diff
@@ -0,0 +1,163 @@
+﻿namespace NServiceBus.Transports.RabbitMQ
+{
+    using Extensibility;
+    using global::RabbitMQ.Client;
+    using global::RabbitMQ.Client.Events;
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Threading.Tasks;
+    using NServiceBus.Logging;
+    using NServiceBus.Transports.RabbitMQ.Config;
+    using NServiceBus.Transports.RabbitMQ.Connection;
+
+    class MessagePump : IPushMessages
+    {
+        readonly ReceiveOptions receiveOptions;
+        ConnectionConfiguration connectionConfiguration;
+        PoisonMessageForwarder poisonMessageForwarder;
+        QueuePurger queuePurger;
+
+        Func<PushContext, Task> pipe;
+        //CriticalError criticalError;
+        PushSettings settings;
+        SecondaryReceiveSettings secondaryReceiveSettings;
+
+        EventingBasicConsumer consumer;
+        TaskCompletionSource<bool> consumerShutdownCompleted;
+        private bool noAck;
+
+        public MessagePump(ReceiveOptions receiveOptions, ConnectionConfiguration connectionConfiguration, PoisonMessageForwarder poisonMessageForwarder, QueuePurger queuePurger)
+        {
+            this.receiveOptions = receiveOptions;
+            this.connectionConfiguration = connectionConfiguration;
+            this.poisonMessageForwarder = poisonMessageForwarder;
+            this.queuePurger = queuePurger;
+        }
+
+        public Task Init(Func<PushContext, Task> pipe, CriticalError criticalError, PushSettings settings)
+        {
+            this.pipe = pipe;
+            //this.criticalError = criticalError;
+            this.settings = settings;
+
+            secondaryReceiveSettings = receiveOptions.GetSettings(settings.InputQueue);
+            noAck = settings.RequiredTransactionMode == TransportTransactionMode.None;
+
+            if (settings.PurgeOnStartup)
+            {
+                queuePurger.Purge(settings.InputQueue);
+            }
+
+            return TaskEx.Completed;
+        }
+
+        public void Start(PushRuntimeSettings limitations)
+        {
+            var taskScheduler = new LimitedConcurrencyLevelTaskScheduler(limitations.MaxConcurrency);
+            var factory = new RabbitMqConnectionFactory(connectionConfiguration, taskScheduler);
+            var connection = new PersistentConnection(factory, connectionConfiguration.RetryDelay, "Consume");
+            var model = connection.CreateModel();
+            model.BasicQos(0, GetPrefetchGount(limitations.MaxConcurrency), false);
+
+            consumer = new EventingBasicConsumer(model);
+            consumerShutdownCompleted = new TaskCompletionSource<bool>();
+
+            consumer.Received += (sender, eventArgs) =>
```

Comment:
Can't find anything, will continue to dig...

The eventing consumer is flagged as experimental

https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/master/projects/client/RabbitMQ.Client/src/client/events/EventingBasicConsumer.cs#L45

are we concerned with that? (perhaps ask the maintainers?)

On Thu, Dec 24, 2015 at 3:56 PM, Brandon Ording <notifications@github.com>
wrote:

> In src/NServiceBus.RabbitMQ/MessagePump.cs
> <https://github.com/Particular/NServiceBus.RabbitMQ/pull/116#discussion_r48417390>
> :
>
> > +
> > +            return TaskEx.Completed;
> > +        }
> > +
> > +        public void Start(PushRuntimeSettings limitations)
> > +        {
> > +            var taskScheduler = new LimitedConcurrencyLevelTaskScheduler(limitations.MaxConcurrency);
> > +            var factory = new RabbitMqConnectionFactory(connectionConfiguration, taskScheduler);
> > +            var connection = new PersistentConnection(factory, connectionConfiguration.RetryDelay, "Consume");
> > +            var model = connection.CreateModel();
> > +            model.BasicQos(0, GetPrefetchGount(limitations.MaxConcurrency), false);
> > +
> > +            consumer = new EventingBasicConsumer(model);
> > +            consumerShutdownCompleted = new TaskCompletionSource<bool>();
> > +
> > +            consumer.Received += (sender, eventArgs) =>
>
> Looking at the implemenation of EventingBasicConsumer, there is nothing
> about how it works that would be different from the DefaultBasicConsumer
> from a concurrency standpoint.
>
> The docs <https://www.rabbitmq.com/dotnet-api-guide.html#concurrency> say
> that "Consumer callbacks are invoked in tasks dispatched to the default
> TaskScheduler provided by the .NET runtime."
>
> By using a custom TaskScheduler instead of the default, we can set a
> maximum concurrency value.
>
> Do you have a link to something that talks about needing multiple
> consumers?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus.RabbitMQ/pull/116/files#r48417390>
> .
>